### PR TITLE
Active Page [manager]: Templates group fix

### PR DESCRIPTION
**Pull request by cannarelladev**
### HOTFIX

[BUG] Active Page (manager): instances are now grouped by its template prettyName (not unique).

## SOLUTION

Now instances are grouped by a unique KEY generated from Template ID + Workspace ID

### CODE REVIEW

File: frontend/src/components/activePage/TableInstance/TableInstance.tsx [hunk 1]
```diff
@@ -54,6 +54,9 @@ const TableInstance: FC<ITableInstanceProps> = ({ ...props }) => {
     });
   };
 
+  const [{ idTemplate, workspaceId }] = instances;
+  const template_key = `${idTemplate}-${workspaceId}`;
```
Comment (QcFe) [inline frontend/src/components/activePage/TableInstance/TableInstance.tsx:58]: Stay camelCase :)

File: frontend/src/utilsLogic.tsx [hunk 2]
```diff
@@ -201,15 +201,16 @@ export const getManagerInstances = (
   index: number
 ) => {
   const { metadata, spec, status } = instance!;
-  const { environmentList, templateName } = spec
-    ?.templateCrownlabsPolitoItTemplateRef?.templateWrapper
+  const { name, namespace, templateWrapper } =
+    spec?.templateCrownlabsPolitoItTemplateRef as any;
```
Comment (QcFe) [inline frontend/src/utilsLogic.tsx:205]: Is the `as any` required?

Comment (QcFe) [reply frontend/src/utilsLogic.tsx:205]: (also below)

File: frontend/src/utilsLogic.tsx [hunk 3]
```diff
@@ -239,38 +240,38 @@ export const getTemplatesMapped = (
...
-    t => {
-      // Find all instances with Template Name === t
-      const instancesFiltered = instances?.filter(
-        ({ templatePrettyName: tpn }) => tpn === t
-      );
-
-      // Find sorting data for Template Name === t
-      const sortDataTmp = sortingData.find(s => s.sortingTemplate === t);
+  return Array.from(
+    new Set(instances?.map(i => `${i.idTemplate}-${i.workspaceId}`))
```
Comment (QcFe) [inline frontend/src/utilsLogic.tsx:244]: Would make a `makeTemplateKey` given an instance, can be reused also below

File: frontend/src/utilsLogic.tsx [hunk 4]
```diff
@@ -239,38 +240,38 @@ export const getTemplatesMapped = (
...
+      instancesFiltered!;
+    return {
+      id: idTemplate,
+      name: templatePrettyName,
+      gui,
+      persistent,
+      resources: { cpu: 0, memory: '', disk: '' },
+      instances: instancesSorted || instancesFiltered,
+      workspaceId,
+    } as Template;
```
Comment (QcFe) [inline frontend/src/utilsLogic.tsx:273]: Is `as Template` really needed?

File: frontend/src/utilsLogic.tsx (hunk 2 above)
Comment (cannarelladev) [reply frontend/src/utilsLogic.tsx:205]: unfortunately yes, for the same ever reason

**Comment by QcFe** [comment]
Something is not working properly. I have a running Instance from a sid template but I can't see it on the workspaces view...

**Comment by cannarelladev** [comment]
> 

what do you mean with workspaces view?

**Comment by QcFe** [comment]
The one on the home page. Also isn't this a duplicate of #697?

**Comment by cannarelladev** [comment]
This PR should fix only the active page, the Dashboard is fixed bya the @cheina97's PR

**Comment by giorio94** [comment]
Converting to draft, due to conflicts

**Comment by QcFe** [comment]
Can we do already here the `type Nullable<T> = T | null | undefind;` thing, for the makeGuiInstance (and template) functions?

**Comment by cannarelladev** [comment]
@QcFe tell me if what i did now is what you intended with previous comment

**Comment by cannarelladev** [comment]
Ready to be merged

File: frontend/src/components/activePage/TableInstance/TableInstanceLogic.tsx [hunk 5]
```diff
@@ -118,7 +118,7 @@ const TableInstanceLogic: FC<ITableInstanceLogicProps> = ({ ...props }) => {
 
   const instances =
     dataInstances?.instanceList?.instances
-      ?.map((i, n) => makeGuiInstance(i ?? {}, tenantId, tenantNamespace))
+      ?.map((i, n) => makeGuiInstance(i ?? {}, tenantId))
```
Comment (QcFe) [inline frontend/src/components/activePage/TableInstance/TableInstanceLogic.tsx:121]:
```suggestion
      ?.map((i, n) => makeGuiInstance(i, tenantId))
```

File: frontend/src/components/workspaces/Templates/TemplatesTableLogic/TemplatesTableLogic.tsx [hunk 6]
```diff
@@ -148,15 +148,10 @@ const TemplatesTableLogic: FC<ITemplateTableLogicProps> = ({ ...props }) => {
...
-              makeGuiInstance(
-                i.data?.createdInstance ?? {},
-                userId ?? '',
-                tenantNamespace,
-                {
-                  templateName: templateId,
-                  workspaceName: workspaceName,
-                }
-              ),
+              makeGuiInstance(i.data?.createdInstance ?? {}, userId ?? '', {
```
Comment (QcFe) [inline frontend/src/components/workspaces/Templates/TemplatesTableLogic/TemplatesTableLogic.tsx:151]:
```suggestion
              makeGuiInstance(i.data?.createdInstance, userId ?? '', {
```

File: frontend/src/components/workspaces/Templates/TemplatesTableLogic/TemplatesTableLogic.tsx [hunk 7]
```diff
@@ -148,15 +148,10 @@ const TemplatesTableLogic: FC<ITemplateTableLogicProps> = ({ ...props }) => {
...
-              makeGuiInstance(
-                i.data?.createdInstance ?? {},
-                userId ?? '',
-                tenantNamespace,
-                {
-                  templateName: templateId,
-                  workspaceName: workspaceName,
-                }
-              ),
+              makeGuiInstance(i.data?.createdInstance, userId ?? '', {
```
Comment (QcFe) [inline frontend/src/components/workspaces/Templates/TemplatesTableLogic/TemplatesTableLogic.tsx:151]:
```suggestion
              makeGuiInstance(i.data?.createdInstance, userId, {
```

File: frontend/src/utilsLogic.tsx [hunk 8]
```diff
@@ -231,43 +241,51 @@ export const joinInstancesAndTemplates = (
...
     gui: guiEnabled,
     persistent: persistent,
-    templatePrettyName: prettyName,
-    templateName: templateName,
-    //TODO: use lables
-    templateId: makeTemplateKey(
-      templateName ?? '',
-      (templateNamespace ?? '').replace(/^workspace-/, '')
-    ),
+    templateId: makeTemplateKey(templateName ?? '', workspaceName),
```
Comment (QcFe) [inline frontend/src/utilsLogic.tsx:278]:
```suggestion
    templateId: makeTemplateKey(templateName, workspaceName),
```

**Review by QcFe — COMMENTED** [review]
Still a couple of `??`s we forgot 🙈

**Review by QcFe — APPROVED** [review]

**Pull request merged**